In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

In [ ]:
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [ ]:
# https://drive.google.com/file/d/1jU38cWxHrKGLg1nWO_wW8031z7d1HpnO/view?usp=drive_link

id='1jU38cWxHrKGLg1nWO_wW8031z7d1HpnO'
downloaded = drive.CreateFile({'id': id})
downloaded.GetContentFile('googleplaystore.csv')

Abrimos el archivo cargando unicamente las columnas necesarias para este ejercicio. La columna Rating no la necesitamos, pero sabemos de otros
ejercicios que hay una fila con un rating imposible, por lo que esa fila hay que eliminarla

In [ ]:
columnas_a_usar = ["App", "Rating", "Genres", "Reviews"]
apps = pd.read_csv("googleplaystore.csv", usecols = columnas_a_usar)

In [ ]:
apps

App  Rating Reviews  \
0         Photo Editor & Candy Camera & Grid & ScrapBook     4.1     159   
1                                    Coloring book moana     3.9     967   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...     4.7   87510   
3                                  Sketch - Draw & Paint     4.5  215644   
4                  Pixel Draw - Number Art Coloring Book     4.3     967   
...                                                  ...     ...     ...   
10836                                   Sya9a Maroc - FR     4.5      38   
10837                   Fr. Mike Schmitz Audio Teachings     5.0       4   
10838                             Parkinson Exercices FR     NaN       3   
10839                      The SCP Foundation DB fr nn5n     4.5     114   
10840      iHoroscope - 2018 Daily Horoscope & Astrology     4.5  398307   

                          Genres  
0                   Art & Design  
1      Art & Design;Pretend Play  
2                   Art & Design  
3                   Art & Design  
4        Art & Design;Creativity  
...                          ...  
10836                  Education  
10837                  Education  
10838                    Medical  
10839          Books & Reference  
10840                  Lifestyle  

[10841 rows x 4 columns]

In [ ]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   App      10841 non-null  object 
 1   Rating   9367 non-null   float64
 2   Reviews  10841 non-null  object 
 3   Genres   10841 non-null  object 
dtypes: float64(1), object(3)
memory usage: 338.9+ KB


Hay aplicaciones repetidas, en mi criterio las sacamos ya que una misma app no puede tener datos diferentes.

In [ ]:
apps["App"].duplicated().sum()

1181

Nos quedamos unicamente con la primer repeticion de aquellas Apps que se repiten

In [ ]:
apps1 = apps.drop_duplicates(subset = "App")
apps1

App  Rating Reviews  \
0         Photo Editor & Candy Camera & Grid & ScrapBook     4.1     159   
1                                    Coloring book moana     3.9     967   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...     4.7   87510   
3                                  Sketch - Draw & Paint     4.5  215644   
4                  Pixel Draw - Number Art Coloring Book     4.3     967   
...                                                  ...     ...     ...   
10836                                   Sya9a Maroc - FR     4.5      38   
10837                   Fr. Mike Schmitz Audio Teachings     5.0       4   
10838                             Parkinson Exercices FR     NaN       3   
10839                      The SCP Foundation DB fr nn5n     4.5     114   
10840      iHoroscope - 2018 Daily Horoscope & Astrology     4.5  398307   

                          Genres  
0                   Art & Design  
1      Art & Design;Pretend Play  
2                   Art & Design  
3                   Art & Design  
4        Art & Design;Creativity  
...                          ...  
10836                  Education  
10837                  Education  
10838                    Medical  
10839          Books & Reference  
10840                  Lifestyle  

[9660 rows x 4 columns]

Los eliminamos correctamente

In [ ]:
apps1["App"].duplicated().sum()

0

In [ ]:
apps1["Rating"].unique()

array([ 4.1,  3.9,  4.7,  4.5,  4.3,  4.4,  3.8,  4.2,  4.6,  3.2,  4. ,
        nan,  4.8,  4.9,  3.6,  3.7,  3.3,  3.4,  3.5,  3.1,  5. ,  2.6,
        3. ,  1.9,  2.5,  2.8,  2.7,  1. ,  2.9,  2.3,  2.2,  1.7,  2. ,
        1.8,  2.4,  1.6,  2.1,  1.4,  1.5,  1.2, 19. ])

Hay un rating que no esta entre 1 y 5, lo eliminamos. Los ratings nans en este ejercicio los dejamos ya que no trabajamos mas con esta columna

In [ ]:
apps2 = apps1.drop(apps1[apps1["Rating"] > 5].index)
apps2["Rating"].unique()

array([4.1, 3.9, 4.7, 4.5, 4.3, 4.4, 3.8, 4.2, 4.6, 3.2, 4. , nan, 4.8,
       4.9, 3.6, 3.7, 3.3, 3.4, 3.5, 3.1, 5. , 2.6, 3. , 1.9, 2.5, 2.8,
       2.7, 1. , 2.9, 2.3, 2.2, 1.7, 2. , 1.8, 2.4, 1.6, 2.1, 1.4, 1.5,
       1.2])

Vemos que hay apps que tienen dos o mas generos. Son aquellas separadas por punto y coma. De paso revisamos que ninguna app tenga dos veces el mismo genero

In [ ]:
apps2["Genres"].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

Por lo que vamos a dejar separados esos generos pero que sigan perteneciendo a la misma app. Dejar los generos todos en minusculas nos asegura
que no haya generos iguales escritos de forma distinta, lo mismo con los espacios. ej: Beauty y beauty no es lo mismo.

In [ ]:
apps2["Genero"] = apps2.Genres.map(lambda x: [genero.lower().strip() for genero in x.split(";")])
apps3 = apps2.explode("Genero")
apps3

App  Rating Reviews  \
0         Photo Editor & Candy Camera & Grid & ScrapBook     4.1     159   
1                                    Coloring book moana     3.9     967   
1                                    Coloring book moana     3.9     967   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...     4.7   87510   
3                                  Sketch - Draw & Paint     4.5  215644   
...                                                  ...     ...     ...   
10836                                   Sya9a Maroc - FR     4.5      38   
10837                   Fr. Mike Schmitz Audio Teachings     5.0       4   
10838                             Parkinson Exercices FR     NaN       3   
10839                      The SCP Foundation DB fr nn5n     4.5     114   
10840      iHoroscope - 2018 Daily Horoscope & Astrology     4.5  398307   

                          Genres             Genero  
0                   Art & Design       art & design  
1      Art & Design;Pretend Play       art & design  
1      Art & Design;Pretend Play       pretend play  
2                   Art & Design       art & design  
3                   Art & Design       art & design  
...                          ...                ...  
10836                  Education          education  
10837                  Education          education  
10838                    Medical            medical  
10839          Books & Reference  books & reference  
10840                  Lifestyle          lifestyle  

[10052 rows x 5 columns]

Nos fijamos que no hayan errores en los generos

In [ ]:
apps3["Genero"].unique()

array(['art & design', 'pretend play', 'creativity', 'action & adventure',
       'auto & vehicles', 'beauty', 'books & reference', 'business',
       'comics', 'communication', 'dating', 'education', 'music & video',
       'brain games', 'entertainment', 'events', 'finance',
       'food & drink', 'health & fitness', 'house & home',
       'libraries & demo', 'lifestyle', 'adventure', 'arcade', 'casual',
       'card', 'action', 'strategy', 'puzzle', 'sports', 'music', 'word',
       'racing', 'simulation', 'board', 'trivia', 'role playing',
       'educational', 'music & audio', 'video players & editors',
       'medical', 'social', 'shopping', 'photography', 'travel & local',
       'tools', 'personalization', 'productivity', 'parenting', 'weather',
       'news & magazines', 'maps & navigation', 'casino'], dtype=object)

Nos fijamos si hay reviews con valor nan

In [ ]:
apps3["Reviews"].isna().sum()

0

Convertimos la columna a int y comprobamos que todos sean numeros positivos

In [ ]:
apps3["Reviews"] = apps3["Reviews"].astype(int)
(apps3["Reviews"] >= 0).sum() == apps3["Reviews"].count()

True

Agrupamos por genero sumando las cantidades de reviews de cada genero. Al haber eliminado al comienzo las apps repetidas, nos aseguramos que
no figuren dos veces las reviews de una misma app. Es cierto que ahora aparecen apps repetidas, pero las reviews van para cada genero de esta app.
ej: Coloring book moana figura dos veces con la misma cantidad de reviews, pero son 967 para Art & Desing y 967 para Pretend Play.

In [ ]:
apps4 = apps3.groupby("Genero").sum()
apps4

<ipython-input-56-b9a4531d5513>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  apps4 = apps3.groupby("Genero").sum()


Rating    Reviews
Genero                                    
action                   1291.3  152110393
action & adventure        411.7   30975158
adventure                 326.9   23048514
arcade                    791.4  119917631
art & design              269.7    1421452
auto & vehicles           305.9    1163666
beauty                    179.7     396240
board                     244.4    2915736
books & reference         742.7   16721794
brain games               274.5    4154042
business                 1077.9    9890245
card                      187.8    7748725
casino                    158.6    5103971
casual                    898.6  144929991
comics                    225.8    2342071
communication            1059.3  285814846
creativity                137.8    1440307
dating                    532.0    3623544
education                2518.6   23761979
educational               381.3    3224638
entertainment            2054.7   58063867
events                    199.6     161018
finance                  1242.9   12662106
food & drink              392.2    6325028
health & fitness         1043.9   21402272
house & home              257.3    1929847
libraries & demo          267.4     906831
lifestyle                1236.4   11835817
maps & navigation         476.3   17729148
medical                  1208.3    1182971
music                      88.6    3963722
music & audio               4.3       1369
music & video             131.1    1514949
news & magazines          840.8   23130228
parenting                 215.0     958331
personalization          1291.0   53543080
photography              1093.4  105351270
pretend play              288.2    4680582
productivity             1259.2   55590649
puzzle                    537.6   26032533
racing                    430.0   55678669
role playing              474.2   22260451
shopping                  761.4   44551730
simulation                830.9   27247409
social                    862.2  227927801
sports                   1140.8   67190447
strategy                  401.6  101974962
tools                    2900.4  229356578
travel & local            761.0   26819741
trivia                    113.1    7175963
video players & editors   606.7   67683966
weather                   305.5   12295164
word                       95.5    5200774

Y por ultimo enseñamos el top 3 de generos con mayor cantidad de reviews

In [ ]:
apps4.nlargest(3, "Reviews")["Reviews"]

Genero
communication    285814846
tools            229356578
social           227927801
Name: Reviews, dtype: int64